### PR TITLE
Client - Event Messages

### PR DESCRIPTION
**Pull request by OSPFNeighbour**
Version 2 of the messaging system I am working on.

Lets call this a POC while I work on making it more modular and testable.

Adds event messages (chat) and allows for system messages to be posted against an event and displayed in the chat.

Allows you to enable/disable notifications for an event. You will automatically subscribe to notifications for an event if you respond positively to the event request. If you are subscribed you will get push notifications for messages

Technically supports images but until we can upload an image to something theres no point in enabling the upload side of the code

This will need a new shell on android for the gradle change for gif support

Added a relationship (as UsersWithEventNotificationEnabled) to event's that holds which users have notifications enabled. This is returned via GQL via a custom function. The function looks for the user in the relationship and returns true or false. 

Did:
- New subscriptions for toggle of notification subs
- Split eventResponse Update into two functions, one just for locations and one for general (makes the logic on sub/unsub easier
- DRY updates to eventResponse updates to collapse heaps of duplicate code
- UI on the client for messages and notification subs
- push emiter for taking event message and working out who has notifications enabled and sending a push
- swapped arg pass on push test function that was backwards of all the others
- Added comments to eventResponse to try make it clearer

TODO: 
- Unread/Read Message smarts so that the messages button can be interactive
- Pagination loading of messages
- Messages UI mutation to pre-load new message on send

screenshot 2018-06-21 14 39 49
screenshot 2018-06-23 13 43 50
screenshot 2018-06-23 13 43 11
screenshot 2018-06-23 13 43 01

### CODE REVIEW

**Comment by OSPFNeighbour** [comment]
screenshot 2018-06-21 14 39 49
screenshot 2018-06-21 14 39 38

File: client/src/screens/messages/messages.js [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+import CREATE_MESSAGE_MUTATION from '../../graphql/create-message.mutation';
+import styles from './styles';
+
+class Messages extends Component {
+  static navigationOptions = {
+    title: 'Event Messages',
+  };
+
+    state = {
```
Comment (sdunster) [inline client/src/screens/messages/messages.js]: indentation

File: server/src/logic.js [hunk 2]
```diff
@@ -27,6 +27,16 @@ const getAuthenticatedDevice = ctx =>
     return device;
   });
 
+const areEventNotificationsEnableForUser = (event, user) =>
+  event.getUsersWithEventNotificationEnabled(
+    { where: { id: user.id } },
+  ).then((result) => {
+    if (result.length) {
```
Comment (sdunster) [inline server/src/logic.js]: just replace these 3 lines with: `return !!result.length`

File: server/src/logic.js [hunk 3]
```diff
@@ -334,7 +344,13 @@ export const getHandlers = ({ models, creators: Creators, push, pubsub }) => {
             user: getAuthenticatedUser(ctx),
             event: Event.findById(args.id),
           })
-          .then(({ event }) => event);
+          .then(({ user, event }) => {
+            const intirimEventObject = event;
```
Comment (sdunster) [inline server/src/logic.js]: s/intirim/interim

File: server/src/logic.js [hunk 4]
```diff
@@ -578,6 +676,31 @@ export const getHandlers = ({ models, creators: Creators, push, pubsub }) => {
           order: [['createdAt', 'DESC']],
         });
       },
+      setEventNotifications(args, ctx) {
+        console.log(args);
```
Comment (sdunster) [inline server/src/logic.js]: remove or make clear what you're logging

File: server/src/logic.js [hunk 5]
```diff
@@ -578,6 +676,31 @@ export const getHandlers = ({ models, creators: Creators, push, pubsub }) => {
...
+        const { eventId, enabled } = args.notifications;
+        getAuthenticatedUser(ctx).then(user =>
```
Comment (sdunster) [inline server/src/logic.js]: I dont see how this can work.. you're not returning the promise.

File: server/src/push/emitters.js [hunk 6]
```diff
@@ -58,6 +59,31 @@ const getPushEmitters = ({ models }) => {
         });
       });
     },
+    pushMessage({ message, eventId }) {
+      if (eventId) {
+        // // TODO: Single joined query
+        Event.findById(eventId).then((event) => {
```
Comment (sdunster) [inline server/src/push/emitters.js]: need to return the promise or use async/await

File: server/tests/message.test.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import { query, itReturnsSuccess } from './common';
```
Comment (sdunster) [inline server/tests/message.test.js]: yay tests

File: server/src/logic.js (hunk 5 above)
Comment (OSPFNeighbour) [reply server/src/logic.js]: I dont need to react to the outcome, I just need it to happen.

File: server/src/push/emitters.js (hunk 6 above)
Comment (OSPFNeighbour) [reply server/src/push/emitters.js]: how come? I dont care about the result I just fire it off and assume it finishes OK

Comment (sdunster) [reply server/src/push/emitters.js]: i believe it's important to present it to the caller. if the caller chooses not to block waiting for it that is fine.

File: server/src/logic.js (hunk 5 above)
Comment (sdunster) [reply server/src/logic.js]: this one you definitely need to fix - the error produced on L685 will never reach the client, otherwise.

**Comment by OSPFNeighbour** [comment]
Aah ok cool. Will fix.

File: server/src/push/emitters.js (hunk 6 above)
Comment (OSPFNeighbour) [reply server/src/push/emitters.js]: none of the other emitters return. should i fix them all in one hit in a seperate PR?

Comment (sdunster) [reply server/src/push/emitters.js]: ideally, yes. fine to leave this as is for now

**Pull request merged**